concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs

```diff
@@ -3467,3 +3482,185 @@ cacheStateAndGetTransactionTable hpbs = do
+                        )
+            -- Insert the account
+            (maybeIndex, nextAccounts) <- Accounts.putNewAccount persistentAccount accounts
+            when (isNothing maybeIndex) $ MTL.throwError "Duplicate account address in genesis accounts."
+            return nextAccounts
+
+        -- Persistent active bakers
+        initialActiveBakersState = (Trie.empty, Trie.empty, 0)
+        updateActiveBakersState (bakers, keys, total) GenesisAccount{gaBaker = Just GenesisBaker{..}} = do
+            newBakers :: BakerIdTrieMap av <- Trie.insert gbBakerId emptyPersistentActiveDelegators bakers
+            newKeys <- Trie.insert gbAggregationVerifyKey () keys
+            let newTotal = total + gbStake
+            return (newBakers, newKeys, newTotal)
+        updateActiveBakersState acc _ = return acc
+        buildActiveBakers (_activeBakers, _aggregationKeys, total) =
+            let _totalActiveCapital = case delegationSupport @av of
+                    SAVDelegationNotSupported -> TotalActiveCapitalV0
+                    SAVDelegationSupported -> TotalActiveCapitalV1 total
+            in  PersistentActiveBakers
+                    { _passiveDelegators = emptyPersistentActiveDelegators,
+                      ..
+                    }
+
+        -- Persistent epoch bakers
+        initialEpochBakersState = (Vec.empty, Vec.empty, 0)
+        updateEpochBakersState (infos, stakes, total) GenesisAccount{gaBaker = Just GenesisBaker{..}} = do
+            let _bieBakerInfo = BaseAccounts.BakerInfo gbBakerId gbElectionVerifyKey gbSignatureVerifyKey gbAggregationVerifyKey
+            let chainParameters = genesisChainParameters :: ChainParameters pv
+            bakerInfoRef <- makePersistentBakerInfoRef @av $
+                case delegationSupport @av of
+                    SAVDelegationNotSupported -> BaseAccounts.BakerInfoExV0 _bieBakerInfo
+                    SAVDelegationSupported ->
+                        let
+                            _bieBakerPoolInfo =
+                                BaseAccounts.BakerPoolInfo
+                                    { _poolOpenStatus = Transactions.OpenForAll,
+                                      _poolMetadataUrl = emptyUrlText,
+                                      _poolCommissionRates = case delegationChainParameters @pv of
+                                        DelegationChainParametersV1 -> chainParameters ^. cpPoolParameters . ppCommissionBounds . to maximumCommissionRates
+                                    }
+                        in
+                            BaseAccounts.BakerInfoExV1{..}
+            return (Vec.snoc infos bakerInfoRef, Vec.snoc stakes gbStake, total + gbStake)
+        updateEpochBakersState acc _ = return acc
+        buildEpochBakers (bakerInfos', bakerStakes', _bakerTotalStake) = do
+            _bakerInfos <- refMake $ BakerInfos bakerInfos'
+            _bakerStakes <- refMake $ BakerStakes bakerStakes'
```

Comment:
It would be ideal if here we immediately flushed data (here and other places above where we are making references) to disk instead of waiting for the end to do that. It saves on traversal.